test/metabase/api/common/internal_test.clj

```diff
@@ -1,9 +1,123 @@
+                       {:status 200 :body body})
+
+(api/malli-defendpoint POST "/post/test-address"
+                       [:as {address :body :as request}]
+                       {address TestAddress}
+                       {:status 200 :body address})
+
+(api/malli-defendpoint POST "/post/closed-test-address"
+                       [:as {address :body :as request}]
+                       {address ClosedTestAddress}
+                       {:status 200 :body address})
+
+(api/define-routes)
+
+(defn- json-mw [handler]
+  (fn [req]
+    (update
+     (handler
+      (update req :body #(-> % slurp (json/parse-string true))))
+     :body json/generate-string)))
+
+(defn exception-mw [handler]
+  (fn [req] (try
+              (handler req)
+              (catch Exception e (mw.exceptions/api-exception-response e)))))
+
+(deftest malli-defendpoint-test
+  (let [server (jetty/run-jetty (json-mw (exception-mw #'routes)) {:port 0 :join? false})
+        port (.. server getURI getPort)
+        post! (fn [route body]
+                (http/post (str "http://localhost:" port route)
+                           {:throw-exceptions false
+                            :accept :json
+                            :as :json
+                            :coerce :always
+                            :body (json/generate-string body)}))]
+    (is (= {:a 1 :b 2} (:body (post! "/post/any" {:a 1 :b 2}))))
+
+    (is (= {:id 1} (:body (post! "/post/id-int" {:id 1}))))
+    (is (= {:errors {:id ["should be an int"]}} (:body (post! "/post/id-int" {:id "1"}))))
+
+    (is (= {:id "myid"
+            :tags ["abc"]
+            :address {:street "abc" :city "sdasd" :zip 2999 :lonlat [0.0 0.0]}}
+           (:body (post! "/post/test-address"
+                         {:id "myid"
+                          :tags ["abc"]
+                          :address {:street "abc"
+                                    :city "sdasd"
+                                    :zip 2999
+                                    :lonlat [0.0 0.0]}}))))
+
+    (is (= {:errors {:address {:id ["missing required key"]
+                               :tags ["missing required key"]
+                               :address ["missing required key"]}}}
+           (:body (post! "/post/test-address" {:x "1"}))))
+
+    (is (= {:errors {:address {:id ["should be a string"]
+                               :tags ["invalid type"]
+                               :address {:street ["missing required key"]
+                                         :zip ["should be an int"]}}}}
+           (:body (post! "/post/test-address" {:id 1288
+                                               :tags "a,b,c"
+                                               :address {:streeqt "abc"
+                                                         :city "sdasd"
+                                                         :zip "12342"
+                                                         :lonlat [0.0 0.0]}}))))
```

Comment:
Here's what an error message looks like

Reply:
also, need to look into how [spell checking](https://github.com/metosin/malli#spell-checking) fits here